src/ldapclient_lib.py

```diff
@@ -0,0 +1,395 @@
+# Copyright 2023 Canonical Ltd.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Library for the ldap-client relation.
+
+This library contains the Provides and Requires classes for handling the ldap-client interface.
+
+
+### Provider Charm
+
+### Requirer Charm
+
+
+"""
+
+import logging
+import pathlib
+import shlex
+import subprocess
+import zipfile
+
+from ops.charm import (
+    CharmBase,
+    CharmEvents,
+    RelationBrokenEvent,
+    RelationChangedEvent,
+    RelationJoinedEvent,
+)
+from ops.framework import EventBase, EventSource, Handle, Object
+from ops.model import ActiveStatus, MaintenanceStatus, ModelError
+
+logger = logging.getLogger(__name__)
+
+"""
+Events - TBD
+"""
+
+# class LdapClientProviderCharmEvents(CharmEvents):
+#    """Events the LDAP Client provider can leverage."""
+# No custom events necessary for the provider, just has to pass along
+# information to the client.
+
+
+class GlauthSnapReadyEvent(EventBase):
+    """Charm Event triggered when a glauth snap is ready to start."""
+
+    def __init__(
+        self,
+        handle: Handle,
+    ):
+        super().__init__(handle)
+
+
+class CertificateAvailableEvent(EventBase):
+    """Charm Event triggered when a CA certificate is available."""
+
+    def __init__(
+        self,
+        handle: Handle,
+        ca_cert: str,
+    ):
+        super().__init__(handle)
+        self.ca_cert = ca_cert
+
+    def snapshot(self) -> dict:
+        """Return snapshot."""
+        return {
+            "ca_cert": self.ca_cert,
+        }
+
+    def restore(self, snapshot: dict):
+        """Restore snapshot."""
+        self.ca_cert = snapshot["ca_cert"]
+
+
+class CertificateUnavailableEvent(EventBase):
+    """Charm Event triggered when a CA certificate is unavailable."""
+
+    pass
```

Comment:
You actually don't need the `pass` here. As long as the docstring is present the Python interpreter is happy:

```suggestion
class CertificateUnavailableEvent(EventBase):
    """Charm Event triggered when a CA certificate is unavailable."""
```

I learned this few weeks ago after I used ellipses `...` all throughout some custom exceptions :laughing: